sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -2785,6 +2794,60 @@ cdef class HaversineDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
 
         return (sin_0 * sin_0 + cos(x1_0) * cos(x2_0) * sin_1 * sin_1)
 
+#------------------------------------------------------------
+# Arccos Distance
+#  D(x, y) = arccos(dot(x, y) / (|x| * |y|)) / PI
+# This is the metric version of cosine distance, i.e. the metric
+# most suitable when you wish to use cosine distance.
+
+cdef class ArccosDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+    cdef inline {{INPUT_DTYPE_t}} cosine_similarity(
+        self,
+        {{INPUT_DTYPE_t}}* x1,
+        {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size
+    ) nogil except -1:
+        cdef {{INPUT_DTYPE_t}} d = ddot(size, x1, 1, x2, 1)
+        cdef {{INPUT_DTYPE_t}} norm1 = ddot(size, x1, 1, x1, 1)
+        cdef {{INPUT_DTYPE_t}} norm2 = ddot(size, x2, 1, x2, 1)
```

Comment:
Can you provide benchmarks for between using BLAS `ddot` and manually performing the dot products, as propose in https://github.com/scikit-learn/scikit-learn/pull/7829#discussion_r87914940?

Reply:
I created a makeshift benchmark in b8837a813 by temporarily adjusting another benchmark, which enabled me to test the performance effects of the subsequent three commits (bd22d28e7, f56fab4f7 and 4deba5b21). I then reverted the makeshift benchmark again in 856342319. Please suggest a neat, permanent way to benchmark this; I am pretty much an outsider.

(Since this was a year ago, I don't remember what came out, although I can probably look it up as I remember that the benchmark suite stores results to disk. In any case, I tend to revert commits that complicate the code without improving performance, so there was probably a benefit.)

Reply:
Hi @jgonggrijp, I am catching up with others topics but I will come back to you shortly. Thanks for your patience.